unison/multierrgroup.go

```diff
@@ -0,0 +1,41 @@
+package unison
+
+import (
+	"context"
+	"sync"
+)
+
+// MultiErrGroup is a collection of goroutines working on subtasks
+// concurrently.  The group waits until all subtasks have finished and collects
+// all errors encountered.
+//
+// The zero value of MultiErrGroup is a valid group.
+type MultiErrGroup struct {
+	mu   sync.Mutex
+	errs []error
+	wg   sync.WaitGroup
+}
+
+// Go starts a new go-routine, collecting errors encounted into the
+// MultiErrGroup.
+func (g *MultiErrGroup) Go(fn func() error) {
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+		err := fn()
+		if err != nil && err != context.Canceled {
```

Comment:
Why would we skip _context.Cancelled_ from being added to the errors list, that could lead to an inconsistent view of the execution and you will need to track it elsewhere?

Reply:
I kinda stumbled on this one. But `Wait` always returns all errors and a common pattern I'm using the group is like this:

```
if errs := grp.Wait(); if len(errs) > 0 {
  // create and return multi-error
}
```

When wrapping (in order to build a tree of errors), I don't really investigate the array, as my aim is to do reporting.

The `context.Canceled` error more or less works like a signal, less than an error. By not filtering out cancellation signaling, users of MultiErrGroup are forced to filter out signaling the moment context.Context is involved.

Alternatively a multierr package that would allow us to do something like this would be nice to have:

```
if errs := multierrs.Ignore(grp.Wait(), context.Canceled); len(errs) > 0 {
  return multierrs.Wrap(errs, "my message")
}
```

Reply:
If one wants to check for cancellation of a context one can do a final check like this:

```
if ctx.Err() ==  context.Cancelled {
  ...
}
```

Reply:
I like the `multierrs.Ignore()` idea you might want to create an issue for that?